module-4/backend/week_four.md

```diff
@@ -5,15 +5,38 @@ Fork this respository. Answer the questions to the best of your ability. Try to
+
 6. Why is continuous integration helpful/important?
+Continuous integration (CI) is the practice of merging all developer working copies to a shared mainline several times a day (like github) and testing every build to make sure it works (like with TravisCI).  For me, as programs get bigger and you are reviewing lots of code, the first question you ask is "does this code even work?".  CI products solve this problem by running the code for you to make sure it works.  When I was jobshadowing with Healthify I asked them about this and they used Travis CI (and it wasn't able to run the app while I was shadowing!).  
+
 7. What are some continuous integration tools?
+TravisCI, CircleCI, AppVeyor, Percy, Buddy, Semaphore.  (I had to google this and had only heard of Travis CI previously).
 
 #### Review  
 
 8. What are some characteristics of "good" git workflow?
+If on a team project not pushing to master and having others review and comment on your code.  Will and I were really good with this on QS, the only issue for us is once things aren't working on Heroku or Github pages then you have to push to master to test your code until it works correctly.  I know for us we have to rebuild the buttons because Github pages wouldn't load our code properly even thought it worked fine on localhost.  Other than that, don't push to master!
+
+I also think using branches, commiting often, having a good commit sytem that works with your project managment app cards,  etc. are all good practices.  
+
 9. What are the four fundamental concepts of object oriented programming?
+ Encapsulation -Tthe internal representation of an object is generally hidden from view outside of the object’s definition.  Programs and components are built out of pieces.
+ Abstraction - Abstracting out elements.  Pieces have limited knowlege of each other.  
+ Inheritance - A way to reuse code of existing objects.  Functionality in layers.   
+ Polymorphism - One name, many forms, ie, more than one identity that fits under a general identity.  
+
 10. What's a module in Ruby and what's the difference between a class and a module?
+Classes are all about objects, modules are all about methods.  Modules are the equliviant of lockers of methods that you can use across multiple classes, like libraries of functionality.  Often, at least at Turing, we use classes as modules (ie lockers of methods) when we really should be using modules.  When I first wrote this answer I used 'functions' instead of 'methods.'  I think JS is runing my Ruby knowlege.  Ha.  
```

Comment:
good explanation. i'd love to hear about use cases for modules. when do you want to use a module over a class?